libvncserver/rfbserver.c

```diff
@@ -2705,6 +2724,74 @@ rfbProcessClientNormalMessage(rfbClientPtr cl)
       }
       return;
 
+    case rfbSetDesktopSize:
+
+        if ((n = rfbReadExact(cl, ((char *)&msg) + 1,
+            sz_rfbSetDesktopSizeMsg - 1)) <= 0) {
+            if (n != 0)
+              rfbLogPerror("rfbProcessClientNormalMessage: read");
+            rfbCloseClient(cl);
+            return;
+        }
+
+        if (msg.sdm.numberOfScreens == 0) {
+            rfbLog("Ignoring setDesktopSize message from client that defines zero screens\n");
+            return;
+        }
+
+        extDesktopScreens = (rfbExtDesktopScreen *) malloc(msg.sdm.numberOfScreens * sz_rfbExtDesktopScreen);
+        if (extDesktopScreens == NULL) {
+                rfbLogPerror("rfbProcessClientNormalMessage: not enough memory");
+                rfbCloseClient(cl);
+                return;
+        }
+
+        if ((n = rfbReadExact(cl, ((char *)extDesktopScreens), msg.sdm.numberOfScreens * sz_rfbExtDesktopScreen)) <= 0) {
+            if (n != 0)
+                rfbLogPerror("rfbProcessClientNormalMessage: read");
+            free(extDesktopScreens);
+            rfbCloseClient(cl);
+            return;
+        }
+        rfbStatRecordMessageRcvd(cl, msg.type, sz_rfbSetDesktopSizeMsg + msg.sdm.numberOfScreens * sz_rfbExtDesktopScreen,
+                                 sz_rfbSetDesktopSizeMsg + msg.sdm.numberOfScreens * sz_rfbExtDesktopScreen);
+
+        for (i=0; i < msg.sdm.numberOfScreens; i++) {
+            extDesktopScreens[i].id = Swap32IfLE(extDesktopScreens[i].id);
+            extDesktopScreens[i].x = Swap16IfLE(extDesktopScreens[i].x);
+            extDesktopScreens[i].y = Swap16IfLE(extDesktopScreens[i].y);
+            extDesktopScreens[i].width = Swap16IfLE(extDesktopScreens[i].width);
+            extDesktopScreens[i].height = Swap16IfLE(extDesktopScreens[i].height);
+            extDesktopScreens[i].flags = Swap32IfLE(extDesktopScreens[i].flags);
+        }
+        msg.sdm.width = Swap16IfLE(msg.sdm.width);
+        msg.sdm.height = Swap16IfLE(msg.sdm.height);
+
+        rfbLog("Client requested resolution change to (%dx%d)\n", msg.sdm.width, msg.sdm.height);
```

Comment:
How do we deal with the first paragraph of https://github.com/rfbproto/rfbproto/blob/master/rfbproto.rst#7410setdesktopsize

> Requests a change of desktop size. This message is an extension and may only be sent if the client has previously received an ExtendedDesktopSize rectangle.

Ignore?